### ISSUE TITLE
Too many modules will be optimized, caching is needed

### ISSUE THREAD

**Issue by torchsmoke** [Nuitka/Nuitka#865]
When I compile my program the modules are optimized first. Unfortunately far too many are optimized. For example PyOpenGL although I import subprocessed. And that leads to the fact that I have to wait half an hour just to optimize all modules.
Solution 1: Nuitka only has to optimize all modules it needs.
Solution 2: Nuitka optimizes them all and stores them in a cache. Then next time only updated modules have to be optimized and saved again and only the needed ones have to be loaded from the cache. This would have the additional advantage that the compilation times are reduced considerably.
What do you think about the solutions?

**Comment by kayhayen** [comment]
The second one is the plan, and exists as an issue already. Without it, Nuitka is practically unusable. Sorry for closing this as a duplicate though.

**Comment by Ben-Faessler** [comment]
If you haven't started yet, I'd like to get started with the caching system. Can I help you?

**Comment by kayhayen** [comment]
I did search a lot now, and shamefully I cannot find it. Many issues for performance mention the topic, but they don't work on solutions.

As far as caching goes, I started with something that #415 would greatly benefit from, namely the writing of used modules to a cache file when doing the demotion. Some modules, which are too large to compile as C code, still take long to compile in Nuitka. And the goal is to do it only once, store it in the cache, and next time we compile, to take the used modules, and create a bytecode module right away with the proper dependencies.

There is `pkg_resources` which includes a copy of pyparsing for whatever reason. And it won't work without it being there, even if rarely ever used. That module takes time to compile each time. Also standard library could be handled as well with full compilation for one time, and properly analyse dependencies, to find that e.g. sqlite is used or not. There is plenty of benefit from just that level of caching.

If you want to peruse that, the bytecode demotion code pointed to in #415 writes things, and the creation of the module from source code, ought to still let plugins have a go (they might modify it), and then then a bytecode module rather than the compile module.

Getting the caching key right will be a large part of the trick. The `sys.path` relevant contents is to be scanned, and I think I started that work. As that affects the used modules analysis. The hash of the sourcecode should be used, and of course Python version, etc.

With that being present, it ought to be possible to demote trouble makers, or whole packages from stdlib that you don't use and have things work fast.

**Comment by Ben-Faessler** [comment]
If I understand that correctly, we need a cache that contains the .o files of the respective scripts and the hash of the source code. So we can actually save the .o for each script and use the hash (sha1?) Of the source code as the file name.

Nuitka would then only have to ask the cache whether the file is available with a certain hash, and if that applies, for example, link this file with a symlink in the build folder.

Is this a good idea?

**Comment by kayhayen** [comment]
When we demote to bytecode, there will be no C code and no object files. 

For tools like `ccache` and `clcache` it's easy to generate the object file from their cache given the same C code is generated. The generation of code doesn't take long. So we never need to care for object files.

Therefore, if compiled modules were to be cached, what is being output when you use `--xml` should be stored and read back, and then have code generated from there, which probably doesn't take long. We could also keep the C code, that is true, and maybe even a better idea. The XML would be more flexible, but the C code has been made independent of other files, so it could in fact be cached instead. But then we need something that is akin to a compiled module, and provides the C source code for compilation (and thus the object file through that other level of caching). But that won't have nodes attached at all.

That approach would limit what plugins can do. I was always thinking, modules should be allowed to also query the node tree of the code in Nuitka. But maybe that's something that would have to be part of the caching key, what input plugins that access the module used. So that wouldn't really be a problem.

I think you made me realize, that I don't need to store and read back XML at all.

Alongside the C code, much like for bytecode demoted stuff, we need to store the used modules of said module, and potentially other things it derives from the code.

Seems like something that is a lot more simple. The hard part about storing and reading back node trees for XML is to get it right such that nothing changed. The hart part about caching is to know if there is not a new plugin that would want to modify the source code, and for it to have the chance. But that's much easier.

**Comment by kayhayen** [comment]
I you can, you ought to start by making the bytecode demotes modules using the information written to the cache already. It should be at least in the current develop version at least. It writes the names of modules only. And it might not have a good key yet, and it's not used to load. That would be a simple change, and start the caching. And later we expand it to ones that are not demoted to bytecode as well by caching the generated C source code too.

In an ideal world, that should allow semi-instant re-compile of even largest programs. Very interesting!

**Comment by Ben-Faessler** [comment]
How can I best test the demoteCompiledModuleToBytecode function? The following case will not happen. 
`        if (
            module.isCompiledPythonModule()
            and module.getCompilationMode() == "bytecode"
        ):
            demoteCompiledModuleToBytecode(module)`

**Comment by kayhayen** [comment]
You want your options to influence `nuitka.tree.Bulding` function `decideCompilationMode`, where this happens. Plugins currently are the only ones to make this happen, but you were to add command line options for it I guess. The top level module cannot be bytecode, but you can hard code things there, although I bet, there might be cases, where plugins need their stuff to be compiled as well, so a global change might not work well.

**Comment by Ben-Faessler** [comment]
I'm having problems following them right now... :(
Can you maybe explain in steps what needs to be done? I would be very grateful.

**Comment by Ben-Faessler** [comment]
I wanted to output things for debugging purposes in the optimize function with optimization_logger.info. However, it does not output anything. Nuitka doesn't even notice if I write logger wrong. Can you please tell me what I am doing wrong?

**Comment by Ben-Faessler** [comment]
If we cache the bytecode, how are they going to reuse it?

**Comment by kayhayen** [comment]
When loading the module, we can do that.

```
            result = CompiledPythonModule(
                module_name=module_name,
                is_top=is_top,
                mode=decideCompilationMode(is_top, module_name, source_ref),
                future_spec=None,
                source_ref=source_ref,
            )

```

Right here, the decided compilation mode is not used immediately. It is after the module is optimized fully, that it will be demoted to a `UncompiledPythonModule` which then is going to lead to including the bytecode only. It then has the used modules attached, so we can recurse into those in optimization passes and code generation. They might not be bytecode of course, and need other modules.

If `mode` was looked at there, and used as a decider to create the compiled or uncompiled module, and the later from cache immediately, then it would not go through optimization as compiled again, once it has been successfully demoted and its used modules are known.

**Comment by Ben-Faessler** [comment]
I have now modified the demoteCompiledModuleToBytecode function so that the bytecode is also saved and loaded. Unfortunately I get an error when executing the executeable
 EOFError: EOF read where object expected

Before I tried to load everything worked.
Here the complete BytecodeDemotion.py (Does not work to paste it as code):
import hashlib
import marshal
import os

from nuitka import Options
from nuitka.importing.ImportCache import (
    isImportedModuleByName,
    replaceImportedModule,
)
from nuitka.importing.Importing import getPackageSearchPath, isPackageDir
from nuitka.ModuleRegistry import replaceRootModule
from nuitka.nodes.ModuleNodes import makeUncompiledPythonModule
from nuitka.plugins.Plugins import Plugins
from nuitka.Tracing import inclusion_logger
from nuitka.utils.AppDirs import getCacheDir
from nuitka.utils.FileOperations import listDir, makePath
from nuitka.utils.Importing import getAllModuleSuffixes


def getModuleImportableFilesHash(full_name):
    package_name = full_name.getPackageName()

    paths = getPackageSearchPath(None)

    if package_name is not None:
        paths += getPackageSearchPath(package_name)

    all_suffixes = getAllModuleSuffixes()

    result_hash = hashlib.md5()

    for count, path in enumerate(paths):
        if not os.path.isdir(path):
            continue

        for fullname, filename in listDir(path):
            if isPackageDir(fullname) or filename.endswith(all_suffixes):
                entry = "%s:%s" % (count, filename)

                if str is not bytes:
                    entry = entry.encode("utf8")

                result_hash.update(entry)

    return result_hash.hexdigest()


def _getCacheFilename(full_name, module_importables_hash, extension):
    module_cache_dir = os.path.join(getCacheDir(), "module-imports-cache")
    makePath(module_cache_dir)

    return os.path.join(
        module_cache_dir, "%s-%s.%s" % (full_name.asString(), module_importables_hash, extension)
    )

def _loadBytecodeFromCache(cache_filename):
    with open(cache_filename, "rb") as bytecode_cache_file:
        return bytecode_cache_file.read()

def demoteCompiledModuleToBytecode(module):
    """Demote a compiled module to uncompiled (bytecode)."""

    full_name = module.getFullName()
    filename = module.getCompileTimeFilename()

    if Options.isShowProgress():
        inclusion_logger.info(
            "Demoting module %r to bytecode from %r." % (full_name, filename)
        )

    module_importables_hash = getModuleImportableFilesHash(full_name)
    if os.path.exists(_getCacheFilename(full_name, module_importables_hash, "txt")):
        inclusion_logger.info("The module %r is already demoted" % (full_name))
        already_demoted = True

    else:
        already_demoted = False
        source_code = module.getSourceCode()

        # Second chance for plugins to modify source code just before turning it
        # to bytecode.
        source_code = Plugins.onFrozenModuleSourceCode(
            module_name=full_name, is_package=False, source_code=source_code
        )

        bytecode = compile(source_code, filename, "exec", dont_inherit=True)

        bytecode = Plugins.onFrozenModuleBytecode(
            module_name=full_name, is_package=False, bytecode=bytecode
        )

    uncompiled_module = makeUncompiledPythonModule(
        module_name=full_name,
        filename=filename,
        bytecode=marshal.dumps(bytecode) if not already_demoted else _loadBytecodeFromCache(_getCacheFilename(full_name, module_importables_hash, "dat")),
        is_package=module.isCompiledPythonPackage(),
        user_provided=True,
        technical=False,
    )

    used_modules = module.getUsedModules()
    print(used_modules)
    uncompiled_module.setUsedModules(used_modules)
    module.finalize()

    if isImportedModuleByName(full_name):
        replaceImportedModule(old=module, new=uncompiled_module)
    replaceRootModule(old=module, new=uncompiled_module)

    from nuitka.plugins.PluginBase import isTriggerModule, replaceTriggerModule

    if isTriggerModule(module):
        replaceTriggerModule(old=module, new=uncompiled_module)

    if not already_demoted:
        with open(_getCacheFilename(full_name, module_importables_hash, "txt"), "w") as modules_cache_file:
            for module_name, _filename in used_modules:
                modules_cache_file.write(module_name.asString() + "\n")

        with open(_getCacheFilename(full_name, module_importables_hash, "dat"), "wb") as bytecode_cache_file:
            bytecode_cache_file.write(marshal.dumps(bytecode))

**Comment by Ben-Faessler** [comment]
I have succeeded in getting the bytecode cached, and loaded if it is already in the cache. What comes next?
For info, I have manipulated the decideCompilationMode so that the result is always "bytecode".

**Comment by kayhayen** [comment]
Adding an option that matches module names that controls it would be nice, something similar to this, the function from modules should lead you down to where the option is to be added.

```
    any_case, reason = module_name.matchesToShellPatterns(
        patterns=Options.getShallFollowModules()
    )
```

The first is a boolean, the second is nice text for tracing. You code would then be equivalent to `*` as pattern.

I think if you import `pkg_resources.extern` you are guarantueed to have a module that is affected without the option, I think `pkg_resources.extern.pyparsing` will be very worth while this treatment. Pyparsing is big and used by nearly all programs, since nearly every library somehow uses `pkg_resources` in some form, and that won't work without its inline copies.

With that, and even without that, feel free to make a pull request.

**Comment by Ben-Faessler** [comment]
Why is the key not the content of the module, but the names of the python files and folders in the same folder as the module? So the cached becomes invalid with every other working folder.

**Comment by kayhayen** [comment]
The optimization is going rely on module presence really soon.

Imagine this:

```
try:
   import PySide2 as Qt
except ImportError:
   import PyQt5 as Qt
```

At the point, bytecode demotion is done, this could be statically optimized already. I will admit that right now it makes no difference, but I think for standalone mode, it's stupid to go through `__import__` built-in and catch exceptions, etc. and soon the code will just know that PyQt5 is not installed, and not list it as a dependency as a result.

The key side of things is something that I will certainly work harder on. Global optimization, which is what this is that I am talking about, could be cached as well as local only. For now, feel free to disable it (make it dependent on a parameter) for now, if it annoys you. But the next level, e.g. using values from imported modules, inlining functions to find out, that some code path is not taken, might result in not including modules, that appear used early on.

I have to admit, this is nothing I have designed too clearly yet.

**Comment by kayhayen** [comment]
Need to at least extract the value from the discussion at some point.

**Comment by woshiagan** [comment]
> 我现在确实搜索了很多，但遗憾的是我找不到它。很多性能问题都提到了这个话题，但是并没有解决方案。
> 
> 就缓存而言，我从#415受益匪浅的事情开始，即在降级时将使用过的模块写入缓存文件。有些模块太大而无法编译为 C 代码，但在 Nuitka 中编译仍然需要很长时间。我们的目标是只执行一次，将其存储在缓存中，下次编译时，获取使用的模块，并立即创建具有适当依赖项的字节码模块。
> 
> 无论出于何种原因，其中`pkg_resources`都包含 pyparsing 的副本。如果没有它，它就无法工作，即使很少使用。该模块每次都需要时间来编译。标准库也可以通过一次完整编译来处理，并正确分析依赖关系，以发现例如 sqlite 是否被使用。这种级别的缓存有很多好处。
> 
> 如果你想仔细阅读，#415中指出的字节码降级代码会写入内容，并且从源代码创建模块，应该仍然让插件可以尝试（他们可能会修改它），然后是字节码模块而不是编译模块。
> 
> 正确获取缓存密钥将是技巧的很大一部分。相关`sys.path`内容要扫描，我想我就开始了这项工作。因为这会影响所使用的模块分析。应该使用源代码的哈希，当然还有Python版本等。
> 
> 有了它，应该可以降级麻烦制造者，或者从 stdlib 中降级您不使用的整个软件包，并让事情快速运行。

Traceback (most recent call last):
  File "F:\nongchao\o\MAIN~1.DIS\main.py", line 4, in <module>
  File "F:\nongchao\o\MAIN~1.DIS\zk.py", line 1, in <module zk>
  File "F:\nongchao\o\main.dist\sqlite3\__init__.py", line 57, in <module>
    from sqlite3.dbapi2 import *
  File "F:\nongchao\o\main.dist\sqlite3\dbapi2.py", line 27, in <module>
    from _sqlite3 import *
ImportError: DLL load failed while importing _sqlite3: 找不到指定的模块。

I encountered this issue while compiling. Is there a way to solve it?

**Comment by KevinRodriguez777** [comment]
> > 我现在确实搜索了很多，但遗憾的是我找不到它。很多性能问题都提到了这个话题，但是并没有解决方案。
> > 就缓存而言，我从#415受益匪浅的事情开始，即在降级时将使用过的模块写入缓存文件。有些模块太大而无法编译为 C 代码，但在 Nuitka 中编译仍然需要很长时间。我们的目标是只执行一次，将其存储在缓存中，下次编译时，获取使用的模块，并立即创建具有适当依赖项的字节码模块。
> > 无论出于何种原因，其中`pkg_resources`都包含 pyparsing 的副本。如果没有它，它就无法工作，即使很少使用。该模块每次都需要时间来编译。标准库也可以通过一次完整编译来处理，并正确分析依赖关系，以发现例如 sqlite 是否被使用。这种级别的缓存有很多好处。
> > 如果你想仔细阅读，#415中指出的字节码降级代码会写入内容，并且从源代码创建模块，应该仍然让插件可以尝试（他们可能会修改它），然后是字节码模块而不是编译模块。
> > 正确获取缓存密钥将是技巧的很大一部分。相关`sys.path`内容要扫描，我想我就开始了这项工作。因为这会影响所使用的模块分析。应该使用源代码的哈希，当然还有Python版本等。
> > 有了它，应该可以降级麻烦制造者，或者从 stdlib 中降级您不使用的整个软件包，并让事情快速运行。
> 
> Traceback (most recent call last): File "F:\nongchao\o\MAIN~1.DIS\main.py", line 4, in File "F:\nongchao\o\MAIN~1.DIS\zk.py", line 1, in File "F:\nongchao\o\main.dist\sqlite3__init__.py", line 57, in from sqlite3.dbapi2 import * File "F:\nongchao\o\main.dist\sqlite3\dbapi2.py", line 27, in from _sqlite3 import * ImportError: DLL load failed while importing _sqlite3: 找不到指定的模块。
> 
> I encountered this issue while compiling. Is there a way to solve it?

open an new issue rather than commenting on a separate issue.

**Comment by kayhayen** [comment]
All of this is implemented by now. The bytecode cache is influenced by plugins that took effect and their inputs, presence modules affects caching too, for the actual XML storage of the --xml output and reading a module back from this, there is nothing in here, so we don't need this issue anyymore.
